storage/storageCacherAdapter/storageCacherAdapter.go

```diff
@@ -0,0 +1,243 @@
+}
+
+// NewStorageCacherAdapter creates a new storageCacherAdapter
+func NewStorageCacherAdapter(
+	cacher storage.AdaptedSizedLRUCache,
+	db storage.Persister,
+	storedDataFactory storage.StoredDataFactory,
+	marshalizer marshal.Marshalizer,
+) (*storageCacherAdapter, error) {
+	if check.IfNil(cacher) {
+		return nil, storage.ErrNilCacher
+	}
+	if check.IfNil(db) {
+		return nil, storage.ErrNilPersister
+	}
+	if check.IfNil(marshalizer) {
+		return nil, storage.ErrNilMarshalizer
+	}
+	if check.IfNil(storedDataFactory) {
+		return nil, storage.ErrNilStoredDataFactory
+	}
+
+	return &storageCacherAdapter{
+		cacher:            cacher,
+		db:                db,
+		lock:              sync.RWMutex{},
+		storedDataFactory: storedDataFactory,
+		marshalizer:       marshalizer,
+	}, nil
+}
+
+// Clear clears the underlying cacher and db
+func (c *storageCacherAdapter) Clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// TODO also clear the underlying db
+	c.cacher.Purge()
+}
+
+// Put adds the given value in the cacher. If the cacher is full, the evicted values will be persisted to the db
+func (c *storageCacherAdapter) Put(key []byte, value interface{}, sizeInBytes int) (evicted bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	evictedValues := c.cacher.AddSizedAndReturnEvicted(string(key), value, int64(sizeInBytes))
+	for evictedKey, evictedVal := range evictedValues {
+		evictedKeyStr, ok := evictedKey.(string)
+		if !ok {
+			log.Warn("invalid key type")
+			continue
+		}
+
+		evictedValBytes, err := c.marshalizer.Marshal(evictedVal)
+		if err != nil {
+			log.Error("could not marshall value", "error", err)
+			continue
+		}
+
+		err = c.db.Put([]byte(evictedKeyStr), evictedValBytes)
+		if err != nil {
+			log.Error("could not save to db", "error", err)
+			continue
+		}
+	}
+
+	return len(evictedValues) != 0
+}
+
+// Get returns the value at the given key
+func (c *storageCacherAdapter) Get(key []byte) (value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	val, ok := c.cacher.Get(string(key))
+	if ok {
+		return val, true
+	}
+
+	valBytes, err := c.db.Get(key)
+	if err != nil {
+		return nil, false
+	}
+
+	storedData := c.storedDataFactory.CreateEmpty()
+	err = c.marshalizer.Unmarshal(storedData, valBytes)
+	if err != nil {
+		log.Error("could not unmarshall", "error", err)
+		return nil, false
+	}
+
+	return storedData, true
+}
+
+// Has checks if the given key is present in the storageUnit
+func (c *storageCacherAdapter) Has(key []byte) bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	isPresent := c.cacher.Contains(string(key))
+	if isPresent {
+		return true
+	}
+
+	err := c.db.Has(key)
+	if err == nil {
+		return true
+	}
+
+	return false
+}
+
+// Peek returns the value at the given key
+func (c *storageCacherAdapter) Peek(key []byte) (value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	return c.cacher.Peek(string(key))
+}
+
+// HasOrAdd adds the given value in the storageUnit
```

Comment:
`// HasOrAdd checks if the value exists and adds it otherwise`

Reply:
Done.